portal/server/helpers/map-facility-properties.ts

```diff
@@ -0,0 +1,38 @@
+import { FacilityDashboard, FACILITY_STAGE } from '@ukef/dtfs2-common';
+
+/**
+ * Maps the facility stage based on the current facility properties.
+ *
+ * @param {Facility} facility - The facility object to map the stage for.
+ * @returns {Facility} The facility object with the mapped stage.
+ *
+ * The function checks the `facilityStage` property of the input `facility`.
+ * If the `facilityStage` is `FACILITY_STAGE.RISK_EXPIRED`, it sets the `facilityStage`
+ * of the returned facility to `FACILITY_STAGE.RISK_EXPIRED`.
+ * Otherwise, it sets the `facilityStage` based on the `hasBeenIssued` property:
+ * - If `hasBeenIssued` is `true`, it sets the `facilityStage` to `FACILITY_STAGE.ISSUED`.
+ * - If `hasBeenIssued` is `false`, it sets the `facilityStage` to `FACILITY_STAGE.UNISSUED`.
+ */
+const mapFacilityStage = (facility: FacilityDashboard): FacilityDashboard => {
+  const mappedFacility = facility;
+
+  if (facility.facilityStage === FACILITY_STAGE.RISK_EXPIRED) {
+    mappedFacility.facilityStage = FACILITY_STAGE.RISK_EXPIRED;
+  } else {
+    mappedFacility.facilityStage = facility.hasBeenIssued ? FACILITY_STAGE.ISSUED : FACILITY_STAGE.UNISSUED;
+  }
+
+  return mappedFacility;
+};
+
+// Mapper accumulator
+const mappings = [mapFacilityStage];
+
+/**
+ * Maps the properties of an array of facilities using a series of mapping functions.
+ *
+ * @param {Facility[]} facilities - The array of facilities to be mapped.
+ * @returns {Facility[]} The array of facilities with mapped properties.
+ */
+export const mapFacilityProperties = (facilities: FacilityDashboard[]): FacilityDashboard[] =>
+  facilities.map((facility) => mappings.reduce((mappedFacilities, mapper) => mapper(mappedFacilities), facility));
```

Comment:
I think combining a `map` and a reduce is a bit confusing. Would this not be simpler

```suggestion
  facilities.map((facility) => {
    const facilityDashboard = cloneDeep(facility);

    facilityDashboard.facilityStage = getFacilityStage(facility);

    return facilityDashboard;
  });
```

with
```
const getFacilityStage = (facility: FacilityDashboard): string => {
  if (facility.facilityStage === FACILITY_STAGE.RISK_EXPIRED) {
    return FACILITY_STAGE.RISK_EXPIRED;
  }

  return facility.hasBeenIssued ? FACILITY_STAGE.ISSUED : FACILITY_STAGE.UNISSUED;
};
```

Reply:
Above snippet does allow for an effecient addition of additional mappers. With current approach all I will have to do is add or remove mappers from an array `const mappings = [mapFacilityStage];`.

We also do not need to deep clone the facility as we should be in position to run multiple mappers in future and save them to an accumulated array.

`reduce` provides an ability to do so in a succint pattern.

Reply:
I see the argument but I still think I'd rather keep it simple unless multiple mappers are needed. Not a big deal though so happy to resolve.

The cloning is necessary though as currently the `facility` object is being mutated I've just checked in a unit test.

![image](https://github.com/user-attachments/assets/c76aff88-9322-41be-b658-6569d303ddc1)